client/backoff/backoff.go

```diff
@@ -68,6 +68,16 @@ func (b *Backoff) Retry(ctx context.Context, f func() error) error {
 	for {
 		err := f()
 		if err != nil {
+			// If the context has errored, return immediately.
+			// Can't just wait for Done() in this case, because
+			// this can race with the backoff, resulting in more
+			// calls to f() than expected.
+			if ctx.Err() != nil {
```

Comment:
I think this was because the initial backoff duration is ridiculously low (1ns). That meant `time.After()` fired practically immediately, and so the select could choose it instead of `ctx.Done()` in some rare cases.